src/qutip_tensorflow/core/data/dense_tensor.py

```diff
@@ -0,0 +1,85 @@
+import qutip as qt
+import numpy as np
+import tensorflow as tf
+import qutip.core.data as data
+import numbers
+
+__all__ = ['DenseTensor']
+
+
+class DenseTensor(data.Data):
+    """This class provide a wraps around TensorFlow's Tensor.
+
+    Parameters
+    ----------
+    data: array-like
+        Data to be stored.
+    shape: (int, int)
+        Shape of data. Default None, it tries to infer the shape from data
+        accessing the attribute `data.shape`.
+    copy: bool
+        Default True. If True creates a copy of the data. If a `tf.Tensor` is
+        provided as data, it will only preserve the graph structure of the
+        original data if copy=False. If anything different to a `tf.Tensor` or
+        `tf.Variable` is provided, it will copy the data regardless of copy as
+        it needs to be moved to the GPU.
+    """
+
+    def __init__(self, data, shape=None, copy=True):
+        # Try to inherit shape from data
+        if shape is None:
+            try:
+                shape = data.shape
+            except AttributeError:
+                raise ValueError("""Shape could not be inferred from data.
+                                 Please, include the shape of data.""")
+
+            if isinstance(shape, tf.TensorShape):
+                shape = tuple(shape.as_list())
+
+            if len(shape) == 0:
+                shape = (1, 1)
+                # Promote to a ket by default if passed 1D data.
+            if len(shape) == 1:
+                shape = (shape[0], 1)
+
+        if not (
+            len(shape) == 2
+            and isinstance(shape[0], numbers.Integral)
+            and isinstance(shape[1], numbers.Integral)
+            and shape[0] > 0
+            and shape[1] > 0
+            and isinstance(shape, tuple)
```

Comment:
This test should be done first. If `shape` is an integer, `len(shape)` will raise an error and you will not get the desired error message.